internal/db/migrate.go

```diff
@@ -10,48 +11,80 @@ import (
 )
 
 func Migrate(connStr string, migrationsDir string, logger *logrus.Logger) error {
-	fmt.Println(migrationsDir, fmt.Sprintf("file://%s", migrationsDir))
-	m, err := migrate.New(fmt.Sprintf("file://%s", migrationsDir), connStr)
+	m, err := prepare(connStr, migrationsDir, logger)
+	defer func() {
+		if m == nil {
+			return
+		}
+		srcErr, dbErr := m.Close()
+		if srcErr != nil {
+			logger.Errorln("Source error when closing migration", srcErr)
+		}
+		if dbErr != nil {
+			logger.Errorln("DB error when closing migration", dbErr)
+		}
+	}()
+	if err != nil {
+		return err
+	}
+	return finish(m, m.Up(), logger)
+}
+
+func MigrateSteps(connStr string, migrationsDir string, steps int, logger *logrus.Logger) error {
+	m, err := prepare(connStr, migrationsDir, logger)
+	defer func() {
```

Comment:
I see that you've extracted these checks from finish, but I don't understand the reason, can you explain me why ?

Reply:
This was actually a small bug, where errors would result in the migration/connection not being closed. I know the placement of the err != nil check and defer might seem a little weird, but the problem is that the migration gets opened in prepare:

If prepare returns an error we need to return early and close the migration, but m might be nil because it could have failed in `migrate.New` already. So we need the nil check in defer.

defer here can't be part of finish because the migration might have been opened in `prepare`, errored out in `prepare` after opening, and it would never reach `finish`, so it would never close, which wouldn't fix the bug.

---

The alternative is moving the migrate.New call into the `Migrate*` functions again, maybe a bit more readable:

```
func MigrateSteps(connStr string, migrationsDir string, steps int, logger *logrus.Logger) error {
	logger.Infoln("Applying migrations from directory ", fmt.Sprintf("file://%s", migrationsDir))
	m, err := migrate.New(fmt.Sprintf("file://%s", migrationsDir), connStr)
	if err != nil {
		return nil, err
	}

	defer func() {
		srcErr, dbErr := m.Close()
		if srcErr != nil {
			logger.Errorln("Source error when closing migration", srcErr)
		}
		if dbErr != nil {
			logger.Errorln("DB error when closing migration", dbErr)
		}
	}()


	m, err := prepare(m, connStr, migrationsDir, logger)
	if err != nil {
		return err
	}
	return finish(m, m.Steps(steps), logger)
}
```

Reply:
I see, that's a good point indeed!

You could also define a new function:
```
func cleanup(m *migrate.Migrate) error {
	if m == nil {
		return
	}
	srcErr, dbErr := m.Close()
	if srcErr != nil {
		logger.Errorln("Source error when closing migration", srcErr)
	}
	if dbErr != nil {
		logger.Errorln("DB error when closing migration", dbErr)
	}
}
```

and call `defer cleanup(m)`